pkg/storage/stores/shipper/indexshipper/gatewayclient/gateway_client.go

```diff
@@ -237,78 +241,183 @@ func (s *GatewayClient) QueryIndex(_ context.Context, _ *logproto.QueryIndexRequ
 	var (
 		resp *logproto.LabelResponse
 		err  error
 	)
 	err = s.poolDo(ctx, func(client logproto.IndexGatewayClient) error {
-		resp, err = client.LabelValuesForMetricName(ctx, in, opts...)
+		resp, err = client.LabelValuesForMetricName(ctx, in)
 		return err
 	})
 	return resp, err
 }
 
-func (s *GatewayClient) GetStats(ctx context.Context, in *logproto.IndexStatsRequest, opts ...grpc.CallOption) (*logproto.IndexStatsResponse, error) {
+func (s *GatewayClient) GetStats(ctx context.Context, in *logproto.IndexStatsRequest) (*logproto.IndexStatsResponse, error) {
 	var (
 		resp *logproto.IndexStatsResponse
 		err  error
 	)
 	err = s.poolDo(ctx, func(client logproto.IndexGatewayClient) error {
-		resp, err = client.GetStats(ctx, in, opts...)
+		resp, err = client.GetStats(ctx, in)
 		return err
 	})
 	return resp, err
 }
 
-func (s *GatewayClient) GetVolume(ctx context.Context, in *logproto.VolumeRequest, opts ...grpc.CallOption) (*logproto.VolumeResponse, error) {
+func (s *GatewayClient) GetVolume(ctx context.Context, in *logproto.VolumeRequest) (*logproto.VolumeResponse, error) {
 	var (
 		resp *logproto.VolumeResponse
 		err  error
 	)
 	err = s.poolDo(ctx, func(client logproto.IndexGatewayClient) error {
-		resp, err = client.GetVolume(ctx, in, opts...)
+		resp, err = client.GetVolume(ctx, in)
 		return err
 	})
 	return resp, err
 }
 
+func (s *GatewayClient) GetShards(
+	ctx context.Context,
+	in *logproto.ShardsRequest,
+) (res *logproto.ShardsResponse, err error) {
+
+	// We try to get the shards from the index gateway,
+	// but if it's not implemented, we fall back to the stats.
+	// We limit the maximum number of errors to 2 to avoid
+	// cascading all requests to new node(s) when
+	// the idx-gw replicas start to update to a version
+	// which supports the new API.
+	var (
+		maxErrs = 2
+		errCt   int
+	)
+
+	if err := s.poolDoWithStrategy(
+		ctx,
+		func(client logproto.IndexGatewayClient) error {
+			perReplicaResult := &logproto.ShardsResponse{}
+			streamer, err := client.GetShards(ctx, in)
+			if err != nil {
+				return errors.Wrap(err, "get shards")
+			}
+
+			// TODO(owen-d): stream currently unused (buffered) because query planning doesn't expect a streamed response,
+			// but can be improved easily in the future by using a stream here.
+			for {
+				resp, err := streamer.Recv()
+				if err == io.EOF {
+					break
+				}
+				if err != nil {
+					return errors.WithStack(err)
+				}
+				perReplicaResult.Shards = append(perReplicaResult.Shards, resp.Shards...)
+			}
+
+			// Since `poolDo` retries on error, we only want to set the response if we got a successful response.
+			// This avoids cases where we add duplicates to the response on retries.
+			res = perReplicaResult
+
+			return nil
+		},
+		func(err error) bool {
+			errCt++
+			return errCt <= maxErrs
+		},
+	); err != nil {
+		if isUnimplementedCallError(err) {
+			return s.getShardsFromStatsFallback(ctx, in)
```

Comment:
This should only happen during a rollout, If some of the index gateways are still running an older image which doesn't implement the GetShards method, right?

If so, for the sake of simplicity, I think we should have a TODO here to clean the old way of computing shards once the rollout is completed.

Reply:
We actually, this is something that would prevent users from rolling out Loki 3.0 w/o panics... Maybe a TODO for a release after 3.0?